### PR TITLE
Implement ci cd

### PR DESCRIPTION
**Pull request by aspirelus**
How to run

For Build Step:
- run command: docker-compose -f Docker-compose.yml up
- open in browser http://localhost:3000/ to check if "you're on rails"

For Deliver Step
- run command: docker-compose -f Docker-compose.yml up
- username and password for login to Jenkins are stored in "ssh_Jenkins" file - the address is https:/localhost:8080/

### CODE REVIEW

File: Docker-compose.yml [hunk 1]
```diff
@@ -32,5 +33,35 @@ services:
       - "3000:3000"
 # The web app can pe accessed through http://localhost:3000/
 
+
+
+
+# Jenkins used for CI part; 
+# Authetication info can be found in ssh_Jenkins file
+  jenkins:
```
Comment (pirvudoru) [inline Docker-compose.yml:41]: I was able to run the docker-compose now. but it seems to me that jenkins does not have the build job. 

I also got a different jenkins password ( `503693001ec1497ba2015e8008198664` ) than** the one listed here: https://github.com/Agilefreaks/devop-test/pull/1/commits/b5a1d1200d1fbdcaa55003a61e3b0588a338d9e9#diff-05c97546509a1856bdc7180d9393f46e3cbdcaa68ac7709c0e5cbf4970a75fbeR4

image

I'm guessing this is not the same for you. Can you figure out why?

File: Dockerfile [hunk 2]
```diff
@@ -25,4 +22,4 @@ ENTRYPOINT ["bin/rails"]
 EXPOSE 3000
 
 # define executable command
-CMD ["rails", "s", "-b", "0.0.0.0"]
\ No newline at end of file
+CMD ["rails", "s","-top", "-b", "0.0.0.0"]
```
Comment (pirvudoru) [inline Dockerfile:25]: Why do you specify both the CMD and the ENTRYPOINT ?

File: ssh_Jenkins [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# Username for authentication in Jenkins
+admin
+# Password for authentication in Jenkins
+790929b5bdde4f9db2133951311f4a16
+
+
+#SSH key for agent credentials in Jenkins
```
Comment (pirvudoru) [inline ssh_Jenkins:7]: This is nice for making the review easier, but I hope you know not to commit secrets in git :) 
Also locally we could do `docker exec -it <container_name or id> bash` to start a terminal in the container.

File: Docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      - POSTGRES_USER=postgres
+      - POSTGRES_PASSWORD=devop_test
+    ports:
+      - "5432:5432"
+    volumes:
+      - dbdata:/var/lib/postgresql/data
+# DB port is always set to 5432
+
+
+  app:
```
Comment (pirvudoru) [inline Docker-compose.yml:17]: When running the provided command ( `docker-compose -f Docker-compose.yml up` ), i'm getting this crash:

```
app_1      | rails aborted!
app_1      | Rake::CommandLineOptionError: Unrecognized --trace option 'op'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:635:in `select_trace_output'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:589:in `block in standard_rake_options'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1623:in `block in parse_in_order'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1577:in `catch'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1577:in `parse_in_order'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1571:in `order!'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1668:in `permute!'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1693:in `parse!'
app_1      | /usr/local/lib/ruby/3.0.0/optparse.rb:1682:in `parse'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:659:in `handle_options'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:92:in `block in init'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:186:in `standard_exception_handling'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/application.rb:89:in `init'
app_1      | /usr/local/bundle/gems/railties-6.1.4.1/lib/rails/commands/rake/rake_command.rb:19:in `block in perform'
app_1      | /usr/local/bundle/gems/rake-13.0.6/lib/rake/rake_module.rb:59:in `with_application'
app_1      | /usr/local/bundle/gems/railties-6.1.4.1/lib/rails/commands/rake/rake_command.rb:18:in `perform'
app_1      | /usr/local/bundle/gems/railties-6.1.4.1/lib/rails/command.rb:50:in `invoke'
app_1      | /usr/local/bundle/gems/railties-6.1.4.1/lib/rails/commands.rb:18:in `<main>'
app_1      | /usr/local/bundle/gems/bootsnap-1.8.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
app_1      | /usr/local/bundle/gems/bootsnap-1.8.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
app_1      | /usr/local/bundle/gems/bootsnap-1.8.0/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
app_1      | /usr/local/bundle/gems/bootsnap-1.8.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
app_1      | /usr/local/bundle/gems/bootsnap-1.8.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
app_1      | /usr/src/app/bin/rails:7:in `<top (required)>'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/client/rails.rb:28:in `load'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/client/rails.rb:28:in `call'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/client/command.rb:7:in `call'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/client.rb:30:in `run'
app_1      | /usr/local/bundle/gems/spring-2.1.1/bin/spring:49:in `<top (required)>'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/binstub.rb:11:in `load'
app_1      | /usr/local/bundle/gems/spring-2.1.1/lib/spring/binstub.rb:11:in `<top (required)>'
app_1      | <internal:/usr/local/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
app_1      | <internal:/usr/local/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
app_1      | /usr/src/app/bin/spring:12:in `block in <top (required)>'
app_1      | <internal:kernel>:90:in `tap'
app_1      | /usr/src/app/bin/spring:9:in `<top (required)>'
app_1      | bin/rails:4:in `load'
app_1      | bin/rails:4:in `<main>'
devop-test_app_1 exited with code 1
```

Is it different on your machine? If yes, why?

File: Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,34 @@
+// there are 4 stages in pipeline, 
+// first is checking the source code from GitHub repo
+// the other three are just displaying text
+
+pipeline {
```
Comment (pirvudoru) [inline Jenkinsfile:5]: Where is this file used?

**Comment by pirvudoru** [comment]
I'm also curious how you did the commits. On https://github.com/Aspirelus/Devop_test_MyVersion I saw commits with the message [Add files via upload](https://github.com/Aspirelus/Devop_test_MyVersion/commit/3ed9195a3626a821713be4608cbe264ddfdfde70)

**Comment by aspirelus** [comment]
I used the command git commit in terminal for some commits and I think for the ones which have the message add via commit, I just uploaded them. That was because I used a VM and sometimes was easier just to upload them. I worked a bit messy, I am not satisfied with that.

**Comment by pirvudoru** [comment]
For some reason the review didn't submit, and haven't noticed it until now. Apologies.

File: Docker-compose.yml (hunk 1 above)
Comment (aspirelus) [reply Docker-compose.yml:41]: I didn't expect that. I checked and apparently my username and password are for my local host and you, using your local host have different credentials. The solution could be, building jenkins on a remote server or using an API token. I generated an API token from my build and if you are logging in with it, I think you should see the build. 
(Access token: 1114ee5273ba6a9d1f270994a0f07a9aad)

File: Dockerfile (hunk 2 above)
Comment (aspirelus) [reply Dockerfile:25]: By mistake. There is no need to specify both in this case, it is enough to use CMD command and delete ENTRYPOINT one.

File: ssh_Jenkins (hunk 3 above)
Comment (aspirelus) [reply ssh_Jenkins:7]: That was my thought, to make the review easier. I know that credentials and sensitive info should be protected but I didn't spend time in that direction during that week. Recently I was looking to git-secrets in solving that.

File: Docker-compose.yml (hunk 4 above)
Comment (aspirelus) [reply Docker-compose.yml:17]: That is happening to me as well now... The problem was that everything worked well until a certain point when started working just "sometimes". I was looking a lot but I can't see the problem, I am missing something...

File: Jenkinsfile (hunk 5 above)
Comment (aspirelus) [reply Jenkinsfile:5]: Jenkinsfile is exported after building the pipeline with the Blue Ocean plugin. Initial I created Jenkinsfile empty, being updated from any change made in pipeline through plugin. This is the pipeline that I built.

**Comment by aspirelus** [comment]
Don't worry, I am sorry too for the late replay. I hope my replays are satisfying your questions.

**Pull request closed without merging**